### PR TITLE
runtime: simplify prepaid gas accounting in view calls

### PR DESCRIPTION
**Pull request by mina86**
Rather than special casing view calls each time prepaid gas is
checked, simply assume that prepaid gas is effectively infinite.  The
max gas burnt limit will still have an effect.

This fixes `sanity/rpc_max_gas_burnt.py` pytest which has been broken
by the recent ‘Simplify gas metering’ commit.  The commit changed
GasLimitExceeded failure into GasExceeded (by changing order in which
error conditions were checked).  This commit changes this back by
effectively disabling the first if branch in `process_gas_limit`
function when performing a view call.

### CODE REVIEW

**Review by bowenwang1996 — COMMENTED** [review]
> The commit changed
GasLimitExceeded failure into GasExceeded (by changing order in which
error conditions were checked)

This is not expected. @olonho @matklad please take a look

**Comment by matklad** [comment]
Good catch, missed this one. Makes me wonder, how many *more* unit-tests are we missing here.

I don't think this is quite the full fix though -- the order is still flipped on non-view calls.

**Comment by mina86** [comment]
True. To be strictly correct we need to check which of the limits – gas limit or prepaid gas – is lower and prefer that. In the edge case where they are equal we just need to pick one or the other. I can prepare change to that effect.

**Comment by mina86** [comment]
<https://github.com/near/nearcore/pull/5187> has the other change.

If general consensus is that this PR is a valid improvement let’s get it merged and we can discuss further on the other PR.

File: runtime/near-vm-logic/src/gas_counter.rs [hunk 1]
```diff
@@ -103,8 +100,7 @@ impl GasCounter {
             self.fast_counter.burnt_gas.checked_add(burn_gas).ok_or(HostError::IntegerOverflow)?;
         let new_used_gas =
             new_burnt_gas.checked_add(new_promises_gas).ok_or(HostError::IntegerOverflow)?;
-        if new_burnt_gas <= self.max_gas_burnt && (self.is_view || new_used_gas <= self.prepaid_gas)
-        {
+        if new_burnt_gas <= self.max_gas_burnt && new_used_gas <= self.prepaid_gas {
```
Comment (mm-near) [inline runtime/near-vm-logic/src/gas_counter.rs:103]: can you add the comment in line 94 - to explain what this method does and what are the arguments?
(what is the difference between 'burn' gas and 'use' gas?)

File: runtime/near-vm-logic/src/gas_counter.rs [hunk 2]
```diff
@@ -266,21 +262,62 @@ impl GasCounter {
...
     #[should_panic]
-    fn test_prepaid_gas_min() {
-        let mut counter = GasCounter::new(ExtCostsConfig::default(), 100, 1, 10, false);
-        counter.deduct_gas(10, 5).unwrap();
+    fn test_burn_gas_must_be_lt_use_gas() {
+        let _ = test_counter(10, 10, false).deduct_gas(5, 2);
+        let _ = test_counter(10, 10, true).deduct_gas(5, 2);
+    }
+
+    #[test]
```
Comment (mm-near) [inline runtime/near-vm-logic/src/gas_counter.rs:287]: thank you for adding the tests !

**Review by mm-near — APPROVED** [review]

**Review by matklad — APPROVED** [review]
Yeah, with tests this is definitely a significant improvement, let's merge.

File: runtime/near-vm-logic/src/gas_counter.rs (hunk 1 above)
Comment (mina86) [reply runtime/near-vm-logic/src/gas_counter.rs:106]: Added some short docs. Problem is that I’m not completely clear on how the gas accounting work.

**Pull request merged**